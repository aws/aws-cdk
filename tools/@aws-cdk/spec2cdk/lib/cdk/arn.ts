import { Resource } from '@aws-cdk/service-spec-types';

/**
 * Find an ARN property for a given resource
 *
 * Returns `undefined` if no ARN property is found, or if the ARN property is already
 * included in the primary identifier.
 */
export function findNonIdentifierArnProperty(resource: Resource) {
  return findArnProperty(resource, (name) => !resource.primaryIdentifier?.includes(name));
}

export function findArnProperty(resource: Resource, filter: (name: string) => boolean = () => true): any {
  const possibleArnNames = ['Arn', `${resource.name}Arn`];
  for (const name of possibleArnNames) {
    const prop = resource.attributes[name];
    if (prop && filter(name)) {
      return name;
    }
  }
  return undefined;
}

