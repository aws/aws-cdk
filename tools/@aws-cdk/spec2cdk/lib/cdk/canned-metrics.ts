import { DimensionSet, Metric, Resource, Service, SpecDatabase } from '@aws-cdk/service-spec-types';
import { ClassType, expr, InterfaceType, IScope, Method, Module, stmt, Type } from '@cdklabs/typewriter';
import {
  metricFunctionName,
  metricsClassNameFromService as metricsClassNameFromNamespace,
} from '../naming/conventions';

/**
 * Generate Canned Metrics
 */
export class CannedMetricsModule extends Module {
  public static forService(db: SpecDatabase, service: Service): CannedMetricsModule {
    const metrics = db.follow('serviceHasMetric', service);
    const namespaces = Array.from(new Set(metrics.map((r) => r.entity.namespace)));
    const cm = new CannedMetricsModule(db, service, namespaces);

    for (const r of metrics) {
      cm.addMetricWithDimensions(r.entity);
    }

    return cm;
  }

  public static forResource(db: SpecDatabase, resource: Resource): CannedMetricsModule {
    const service = db.incoming('hasResource', resource).only().entity;
    const metrics = db.follow('resourceHasMetric', resource);
    const namespaces = Array.from(new Set(metrics.map((r) => r.entity.namespace)));
    const cm = new CannedMetricsModule(db, service, namespaces);

    for (const r of metrics) {
      cm.addMetricWithDimensions(r.entity);
    }

    return cm;
  }

  private metrics: Record<string, MetricsClass> = {};
  private _hasCannedMetrics: boolean = false;

  private constructor(private readonly db: SpecDatabase, service: Service, namespaces: string[]) {
    super(`${service.name}.canned-metrics`);

    const returnType = new MetricsReturnType(this);

    for (const namespace of namespaces) {
      this.metrics[namespace] = new MetricsClass(this, namespace, returnType);
    }
  }

  public get hasCannedMetrics() {
    return this._hasCannedMetrics;
  }

  /**
   * Add metrics for a given dimension set to the module
   */
  public addMetricWithDimensions(metric: Metric) {
    this._hasCannedMetrics = true;
    const dimensions = this.db.follow('usesDimensionSet', metric).map((m) => m.entity);
    this.metrics[metric.namespace].addMetricWithDimensions(metric, dimensions);
  }
}

export class MetricsReturnType extends InterfaceType {
  public constructor(scope: IScope) {
    super(scope, {
      name: 'MetricWithDims',
      export: true,
      properties: [
        {
          name: 'namespace',
          type: Type.STRING,
          immutable: true,
        },
        {
          name: 'metricName',
          type: Type.STRING,
          immutable: true,
        },
        {
          name: 'statistic',
          type: Type.STRING,
          immutable: true,
        },
      ],
    });

    const D = this.addTypeParameter({ name: 'D' });

    this.addProperty({
      name: 'dimensionsMap',
      type: D.asType(),
      immutable: true,
    });
  }
}

export class MetricsClass extends ClassType {
  constructor(scope: IScope, namespace: string, private returnType: MetricsReturnType) {
    super(scope, {
      export: true,
      name: metricsClassNameFromNamespace(namespace),
    });
  }

  public addMetricWithDimensions(metric: Metric, dimensionSets: DimensionSet[]) {
    const name = metricFunctionName(metric);

    // Add a unique declaration for each dimension set
    for (const set of dimensionSets) {
      const dimensionsType = dimensionSetType(set);
      this.addMetricMethodDeclaration(name, dimensionsType);
    }

    // If we have more than one dimension set, add a generic declaration
    if (dimensionSets.length > 1) {
      this.addMetricMethodDeclaration(name, Type.ANY);
    }

    // Add the implementation to the final declaration
    this.methods.at(-1)?.addBody(
      stmt.ret(
        expr.object({
          namespace: expr.lit(metric.namespace),
          metricName: expr.lit(metric.name),
          dimensionsMap: expr.ident('dimensions'),
          statistic: expr.lit(metric.statistic),
        }),
      ),
    );
  }

  private addMetricMethodDeclaration(name: string, dimensionsType: Type): Method {
    return this.addMethod({
      name,
      static: true,
      returnType: Type.fromName(this.scope, this.returnType.name, [dimensionsType]),
      parameters: [
        {
          name: 'dimensions',
          type: dimensionsType,
        },
      ],
    });
  }
}

function dimensionSetType(set: DimensionSet): Type {
  return Type.anonymousInterface(set.dimensions.map(({ name }) => ({ name, type: Type.STRING })));
}
