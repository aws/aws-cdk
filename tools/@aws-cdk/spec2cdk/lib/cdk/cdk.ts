import { $E, $T, expr, Expression, ExternalModule, IScope, ThingSymbol, Type } from '@cdklabs/typewriter';

export interface ModuleImportLocations {
  /**
   * The import name used import the core module
   * @default 'aws-cdk-lib'
   */
  readonly core?: string;
  /**
   * The import name used to import core helpers module
   * @default 'aws-cdk-lib/core/lib/helpers-internal'
   */
  readonly coreHelpers?: string;

  /**
   * The import name used to import the CloudWatch module
   *
   * @default 'aws-cdk-lib/aws-cloudwatch'
   */
  readonly cloudwatch?: string;
}

export class CdkCore extends ExternalModule {
  public readonly helpers = new CdkInternalHelpers(this);

  public readonly CfnResource = Type.fromName(this, 'CfnResource');
  public readonly Resource = $T(Type.fromName(this, 'Resource'));
  public readonly IInspectable = Type.fromName(this, 'IInspectable');
  public readonly TreeInspector = Type.fromName(this, 'TreeInspector');
  public readonly Token = $T(Type.fromName(this, 'Token'));
  public readonly ResolutionTypeHint = Type.fromName(this, 'ResolutionTypeHint');
  public readonly CfnTag = Type.fromName(this, 'CfnTag');
  public readonly TagManager = $T(Type.fromName(this, 'TagManager'));
  public readonly TagType = $T(Type.fromName(this, 'TagType'));
  public readonly ITaggable = Type.fromName(this, 'ITaggable');
  public readonly ITaggableV2 = Type.fromName(this, 'ITaggableV2');
  public readonly IResolvable = Type.fromName(this, 'IResolvable');

  public readonly objectToCloudFormation = makeCallableExpr(this, 'objectToCloudFormation');
  public readonly stringToCloudFormation = makeCallableExpr(this, 'stringToCloudFormation');
  public readonly dateToCloudFormation = makeCallableExpr(this, 'dateToCloudFormation');
  public readonly booleanToCloudFormation = makeCallableExpr(this, 'booleanToCloudFormation');
  public readonly numberToCloudFormation = makeCallableExpr(this, 'numberToCloudFormation');
  public readonly cfnTagToCloudFormation = makeCallableExpr(this, 'cfnTagToCloudFormation');
  public readonly canInspect = makeCallableExpr(this, 'canInspect');
  public readonly listMapper = makeCallableExpr(this, 'listMapper');
  public readonly hashMapper = makeCallableExpr(this, 'hashMapper');
  public readonly unionMapper = makeCallableExpr(this, 'unionMapper');
  public readonly requireProperty = makeCallableExpr(this, 'requireProperty');
  public readonly isResolvableObject = makeCallableExpr(this, 'isResolvableObject');

  public readonly ValidationResult = $T(Type.fromName(this, 'ValidationResult'));
  public readonly VALIDATION_SUCCESS = makeCallableExpr(this, 'VALIDATION_SUCCESS');
  public readonly ValidationResults = $T(Type.fromName(this, 'ValidationResults'));

  public readonly propertyValidator = makeCallableExpr(this, 'propertyValidator');
  public readonly requiredValidator = makeCallableExpr(this, 'requiredValidator');
  public readonly listValidator = makeCallableExpr(this, 'listValidator');
  public readonly hashValidator = makeCallableExpr(this, 'hashValidator');
  public readonly unionValidator = makeCallableExpr(this, 'unionValidator');
  public readonly validateCfnTag = makeCallableExpr(this, 'validateCfnTag');
  public readonly validateObject = makeCallableExpr(this, 'validateObject');
  public readonly validateDate = makeCallableExpr(this, 'validateDate');
  public readonly validateBoolean = makeCallableExpr(this, 'validateBoolean');
  public readonly validateNumber = makeCallableExpr(this, 'validateNumber');
  public readonly validateString = makeCallableExpr(this, 'validateString');

  constructor(fqn: string) {
    super(fqn);
  }

  public tokenAsString(arg: Expression) {
    return this.Token.asString(arg);
  }

  public tokenAsNumber(arg: Expression) {
    return this.Token.asNumber(arg);
  }

  public tokenAsList(arg: Expression) {
    return this.Token.asList(arg);
  }
}

export class CdkInternalHelpers extends ExternalModule {
  public readonly FromCloudFormationOptions = Type.fromName(this, 'FromCloudFormationOptions');
  public readonly FromCloudFormationResult = $T(Type.fromName(this, 'FromCloudFormationResult'));
  public readonly FromCloudFormation = $T(Type.fromName(this, 'FromCloudFormation'));
  public readonly FromCloudFormationPropertyObject = Type.fromName(this, 'FromCloudFormationPropertyObject');

  constructor(parent: CdkCore) {
    super(`${parent.fqn}/core/lib/helpers-internal`);
  }
}

export class Constructs extends ExternalModule {
  public readonly Construct = Type.fromName(this, 'Construct');
  public readonly IConstruct = Type.fromName(this, 'IConstruct');

  constructor() {
    super('constructs');
  }
}

export class CdkCloudWatch extends ExternalModule {
  public readonly Metric = $T(Type.fromName(this, 'Metric'));
  public readonly MetricOptions = Type.fromName(this, 'MetricOptions');
}

export const CDK_CORE = new CdkCore('aws-cdk-lib');
export const CDK_CLOUDWATCH = new CdkCloudWatch('aws-cdk-lib/aws-cloudwatch');
export const CONSTRUCTS = new Constructs();

function makeCallableExpr(scope: IScope, name: string) {
  return $E(expr.sym(new ThingSymbol(name, scope)));
}
