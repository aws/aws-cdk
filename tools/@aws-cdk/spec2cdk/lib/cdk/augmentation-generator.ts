import { Resource, ResourceAugmentation, ResourceMetric, SpecDatabase } from '@aws-cdk/service-spec-types';
import {
  $E,
  ClassType,
  expr,
  InterfaceType,
  Module,
  MonkeyPatchedType,
  Splat,
  MemberType,
  stmt,
  Type,
} from '@cdklabs/typewriter';
import { CDK_CLOUDWATCH, CONSTRUCTS } from './cdk';
import { ResourceClass } from './resource-class';

/**
 * Generate augmentation methods for the given types
 *
 * Augmentation consists of two parts:
 *
 * - Adding method declarations to an interface (IBucket)
 * - Adding implementations for those methods to the base class (BucketBase)
 *
 * The augmentation file must be imported in `index.ts`.
 *
 * ----------------------------------------------------------
 *
 * Generates code similar to the following:
 *
 * ```
 * import <Class>Base from './<class>-base';
 *
 * declare module './<class>-base' {
 *   interface <IClass> {
 *     method(...): Type;
 *   }
 *   interface <ClassBase> {
 *     method(...): Type;
 *   }
 * }
 *
 * <ClassBase>.prototype.<method> = // ...impl...
 * ```
 *
 * This code may not have been factored the best in terms of how it should
 * be modeled in typewriter.
 */
export class AugmentationsModule extends Module {
  private _hasAugmentations: boolean = false;

  /**
   * Modules that contain classes that are normally handwritten
   */
  public readonly supportModules = new Array<Module>();

  constructor(private readonly db: SpecDatabase, serviceName: string, cloudWatchModuleImport?: string) {
    super(`${serviceName}.augmentations`);

    this.documentation.push(
      `Copyright 2012-${new Date().getFullYear()} Amazon.com, Inc. or its affiliates. All Rights Reserved.`,
    );

    CDK_CLOUDWATCH.import(this, 'cw', {
      fromLocation: cloudWatchModuleImport,
    });
  }

  public get hasAugmentations() {
    return this._hasAugmentations;
  }

  public augmentResource(resource: Resource, resourceClass: ResourceClass) {
    for (const { entity: aug } of this.db.follow('isAugmented', resource)) {
      if (aug.metrics) {
        this._hasAugmentations = true;
        new ResourceGenerator(resource, resourceClass, aug, this.supportModules).emit(this);
      }
    }
  }
}

class ResourceGenerator {
  private readonly interfaceFile: string;
  private readonly classFile: string;
  private readonly interfaceName: string;
  private readonly className: string;

  constructor(
    private readonly resource: Resource,
    resourceClass: ResourceClass,
    private readonly aug: ResourceAugmentation,
    private readonly supportModules: Module[],
  ) {
    const l2ClassName = resourceClass.name.replace(/^Cfn/, '');
    const l2KebabName = l2ClassName.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();

    this.classFile = `./${aug.baseClassFile ?? `${l2KebabName}-base`}`;
    this.className = aug.baseClass ?? `${l2ClassName}Base`;
    this.interfaceFile = aug.interfaceFile ? `./${aug.interfaceFile}` : this.classFile;
    this.interfaceName = aug.interface ?? `I${l2ClassName}`;
  }

  public emit(into: Module) {
    const interfaceModule = new Module(this.interfaceFile);
    const classModule = this.classFile === this.interfaceFile ? interfaceModule : new Module(this.classFile);
    this.supportModules.push(...new Set([interfaceModule, classModule]));

    classModule.importSelective(into, [this.className]);

    const externalInterfaceType = new InterfaceType(interfaceModule, { name: this.interfaceName, export: true });
    const externalClassType = this.generateClassType(classModule);

    this.emitPatches(into, externalInterfaceType);
    this.emitPatches(into, externalClassType);
  }

  /**
   * Generate a ClassType representing the L2 class that we're augmenting
   *
   * We didn't need this class at all, in principle, but if we want to be able to
   * generate the code and compile it independent of the human-written code we're
   * going to integrate it into later, we need a representation of this class with
   * the right attributes so that the compiler can do some sensible type checking.
   *
   * This class will only be written to disk if that is explicitly requested during
   * code generation.
   */
  private generateClassType(classModule: Module) {
    CONSTRUCTS.importSelective(classModule, ['Construct']);

    const type = new ClassType(classModule, {
      name: this.className,
      export: true,
      extends: CONSTRUCTS.Construct,
    });
    for (const attrName of Object.values(this.aug.metrics?.dimensions ?? {})) {
      type.addProperty({
        name: attrName,
        type: Type.STRING,
        immutable: true,
        initializer: expr.lit('dummy'),
      });
    }

    return type;
  }

  /**
   * Emit the interface declarations of our mixins
   *
   * The declarations will be emitted as an interface, and TypeScript will
   * combine them with existing declarations, of either an interface or class type.
   */
  private emitPatches(into: Module, targetType: MemberType) {
    const iface = new MonkeyPatchedType(into, targetType);
    this.emitGenericMethod(iface);

    for (const metric of this.aug.metrics?.metrics ?? []) {
      this.emitSpecificMethod(iface, metric);
    }
  }

  private emitGenericMethod(type: MemberType): void {
    const meth = type.addMethod({
      name: 'metric',
      docs: {
        summary: `Return the given named metric for this ${this.resource.name}`,
      },
      returnType: CDK_CLOUDWATCH.Metric,
    });

    const metricName = meth.addParameter({ name: 'metricName', type: Type.STRING });
    const props = meth.addParameter({ name: 'props', type: CDK_CLOUDWATCH.MetricOptions, optional: true });

    const $this = $E(expr.this_());

    meth.addBody(
      stmt.ret(
        new CDK_CLOUDWATCH.Metric(
          expr.object(
            {
              namespace: expr.lit(this.aug.metrics?.namespace),
              metricName,
              dimensionsMap: expr.object(
                Object.entries(this.aug.metrics?.dimensions ?? {}).map(
                  ([name, attrName]) => [name, $this[attrName]] as const,
                ),
              ),
            },
            new Splat(props),
          ),
        ).attachTo($this),
      ),
    );
  }

  private emitSpecificMethod(iface: MemberType, metric: ResourceMetric) {
    const meth = iface.addMethod({
      name: metricFunctionName(metric),
      docs: {
        summary: metric.documentation,
        remarks: `${metricStatistic(metric)} over 5 minutes`,
      },
      returnType: CDK_CLOUDWATCH.Metric,
    });

    const props = meth.addParameter({
      name: 'props',
      type: CDK_CLOUDWATCH.MetricOptions,
      optional: true,
    });

    const $this = $E(expr.this_());
    meth.addBody(
      stmt.ret(
        $this.metric(
          expr.lit(metric.name),
          expr.object(
            {
              statistic: expr.lit(metricStatistic(metric)),
            },
            new Splat(props),
          ),
        ),
      ),
    );
  }
}

function metricFunctionName(metric: ResourceMetric): string {
  return `metric${metric.name.replace(/[^a-zA-Z0-9]/g, '')}`;
}

function metricStatistic(metric: ResourceMetric): string {
  switch (metric.type) {
    case 'attrib':
    case undefined:
      return 'Average';

    case 'count':
      return 'Sum';

    case 'gauge':
      return 'Maximum';
  }
}
