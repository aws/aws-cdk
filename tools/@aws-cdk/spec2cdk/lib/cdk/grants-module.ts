import { promises as fs } from 'fs';
import * as path from 'node:path';
import { Resource, Service, SpecDatabase } from '@aws-cdk/service-spec-types';
import {
  $E,
  ClassType,
  expr,
  Expression,
  ExternalModule,
  InterfaceType,
  MemberVisibility,
  Module,
  stmt,
  Type,
} from '@cdklabs/typewriter';
import { PropertySpec } from '@cdklabs/typewriter/lib/property';
import { camelcasedResourceName, referencePropertyName } from '../naming';
import { findArnProperty } from './arn';

const $this = $E(expr.this_());

/**
 * From some data, generate grants
 *
 * For now, the grants data will be in a JSON file inside the service directory;
 * in the future, this data will be moved to the `awscdk-service-spec`
 * repository.
 */
export class GrantsModule extends Module {
  public static async forServiceFromString(db: SpecDatabase, service: Service, config: string): Promise<GrantsModule | undefined> {
    const schema = JSON.parse(config);
    return new GrantsModule(service, db, schema);
  }

  public static async forServiceFromFile(db: SpecDatabase, service: Service, sourceFile: string): Promise<GrantsModule | undefined> {
    try {
      const location = path.join(__dirname, '../../../../../packages/aws-cdk-lib', sourceFile);
      return await GrantsModule.forServiceFromString(db, service, await fs.readFile(location, 'utf-8'));
    } catch (e: any) {
      if (e.code === 'ENOENT') {
        return undefined;
      }
      throw e;
    }
  }

  private constructor(private readonly service: Service, private readonly db: SpecDatabase, private readonly schema: GrantsFileSchema) {
    super(`${service.shortName}.grants`);
    this.build();
  }

  private build() {
    let hasContent = false;
    const resources = this.db.follow('hasResource', this.service);
    const resourceIndex = Object.fromEntries(resources.map(r => [r.entity.name, r.entity]));
    const encryptedResourcePropName = 'encryptedResource';

    for (const [name, config] of Object.entries(this.schema.resources ?? {})) {
      if (!resourceIndex[name]) continue;
      const resource = resourceIndex[name];

      const hasPolicy = config.hasResourcePolicy ?? false;
      const hasKeyActions = Object.values(config.grants)
        .some(grant => grant.keyActions && grant.keyActions.length > 0);

      const arnMap = resourceArnMap(resource, config.grants);
      if (Object.keys(arnMap).length === 0) continue;

      const className = `${name}Grants`;

      const propsProperties: PropertySpec[] = [{
        name: 'resource',
        type: Type.fromName(this, `${this.service.shortName}.I${resource.name}Ref`),
        immutable: true,
      }];

      if (hasPolicy) {
        propsProperties.push({
          name: 'policyResource',
          type: Type.fromName(this, 'iam.IResourceWithPolicy'),
          optional: true,
          immutable: true,
        });
      }
      if (hasKeyActions) {
        propsProperties.push({
          name: encryptedResourcePropName,
          type: Type.fromName(this, 'iam.IEncryptedResource'),
          optional: true,
          immutable: true,
        });
      }

      const propsType = new InterfaceType(this, {
        name: `${className}Props`,
        export: false,
        properties: propsProperties,
      });

      const classType = new ClassType(this, {
        name: className,
        export: true,
      });

      // IBucketRef
      // FIXME: Not sure that Type.fromName() is the best to be used here.
      const resourceRefType = Type.fromName(this, `${this.service.shortName}.I${resource.name}Ref`);

      classType.addProperty({
        name: 'resource',
        immutable: true,
        type: resourceRefType,
      });

      const init = classType.addInitializer({
        // FIXME this is not having any effect
        visibility: MemberVisibility.Private,
      });

      const propsParameter = init.addParameter({
        name: 'props',
        type: propsType.type,
      });

      init.addBody(stmt.assign($this.resource, propsParameter.prop('resource')));

      if (hasKeyActions) {
        const iEncryptedResourceType = Type.fromName(this, 'iam.IEncryptedResource');
        classType.addProperty({
          name: encryptedResourcePropName,
          immutable: true,
          type: iEncryptedResourceType,
          optional: true,
        });

        init.addBody(stmt.assign($this[encryptedResourcePropName], propsParameter.prop(encryptedResourcePropName)));
      }

      if (hasPolicy) {
        const resourceWithPolicy = Type.fromName(this, 'iam.IResourceWithPolicy');
        classType.addProperty({
          name: 'policyResource',
          immutable: true,
          type: resourceWithPolicy,
          optional: true,
        });
        init.addBody(stmt.assign($this.policyResource, propsParameter.prop('policyResource')));
      }

      const factoryMethod = classType.addMethod({
        name: `from${resource.name}`,
        static: true,
        returnType: Type.fromName(this, `${classType.name}`),
      });

      const staticResourceParam = factoryMethod.addParameter({
        name: 'resource',
        type: resourceRefType,
      });

      const props: Record<string, Expression> = {
        resource: staticResourceParam,
      };

      if (hasKeyActions) {
        props[encryptedResourcePropName] = expr.cond(
          $E(expr.ident('iam.GrantableResources')).isEncryptedResource(staticResourceParam),
          staticResourceParam,
          expr.UNDEFINED,
        );
      }

      if (hasPolicy) {
        props.policyResource = expr.cond(
          $E(expr.ident('iam.GrantableResources')).isResourceWithPolicy(staticResourceParam),
          staticResourceParam,
          expr.UNDEFINED,
        );
      }

      factoryMethod.addBody(
        stmt.ret(Type.fromName(this, className).newInstance(expr.object(props))),
      );

      for (const [methodName, grantSchema] of Object.entries(config.grants)) {
        const resourceArns = arnMap[methodName];
        if (resourceArns == null) {
          // Without ARNs, we cannot generate grants.
          continue;
        }

        const method = classType.addMethod({
          name: methodName,
          returnType: Type.fromName(this, 'iam.Grant'),
        });

        const grantee = method.addParameter({
          name: 'grantee',
          type: Type.fromName(this, 'iam.IGrantable'),
        });

        const actions = expr.ident('actions');

        const addToPrincipalExpr = $E(expr.ident('iam.Grant')).addToPrincipal(expr.object({
          actions,
          grantee,
          resourceArns,
        }));
        const addToBothExpr = $E(expr.ident('iam.Grant')).addToPrincipalOrResource(expr.object({
          actions,
          grantee,
          resourceArns,
          resource: $this.policyResource,
        }));

        method.addBody(stmt.constVar(actions, expr.lit(grantSchema.actions)));

        if (grantSchema && grantSchema.keyActions && grantSchema.keyActions.length > 0) {
          const grantOnKey = $this.prop(`${encryptedResourcePropName}?`)
            .callMethod('grantOnKey', grantee, ...grantSchema.keyActions.map((a) => expr.lit(a)));

          method.addBody(grantOnKey);
        }

        if (hasPolicy) {
          method.addBody(stmt.ret(expr.cond($this.policyResource, addToBothExpr, addToPrincipalExpr)));
        } else {
          method.addBody(stmt.ret(addToPrincipalExpr));
        }

        hasContent = true;
      }
    }

    if (hasContent) {
      new ExternalModule(`aws-cdk-lib/aws-${this.service.shortName}`).import(this, this.service.shortName);
      new ExternalModule('aws-cdk-lib/aws-iam').import(this, 'iam');
    }
  }
}

export interface GrantsFileSchema {
  readonly resources: Record<string, ResourceSchema>;
  readonly constants?: Record<string, string[]>;
}

export interface ResourceSchema {
  readonly hasResourcePolicy?: boolean;
  readonly grants: Record<string, GrantSchema>;
}

export interface GrantSchema {
  /**
   * ARN format containing placeholders
   *
   * If absent, just use the resource's default ARN format.
   */
  readonly arnFormat?: string;
  readonly actions: string[];
  readonly keyActions?: string[];
}

function resourceArnMap(resource: Resource, grants: Record<string, GrantSchema>): Record<string, Expression> {
  const result: Record<string, Expression> = {};
  for (const [methodName, grantSchema] of Object.entries(grants)) {
    let expression = arns(resource, grantSchema);
    if (expression != null) {
      result[methodName] = expression;
    }
  }
  return result;
}

function arns(resource: Resource, grantSchema?: GrantSchema): Expression | undefined {
  const arnFormat = grantSchema?.arnFormat;
  const refName = `this.resource.${camelcasedResourceName(resource)}Ref`;

  if (arnFormat != null) {
    return expr.list([expr.strConcat(...replace(refName, arnFormat))]);
  }

  const cfnArnProperty = findArnProperty(resource);
  return cfnArnProperty == null
    ? undefined
    : expr.list([expr.ident(`${refName}.${referencePropertyName(cfnArnProperty, resource.name)}`)]);
}

// TODO better name
function replace(refName: string, arnFormat: string): Expression[] {
  const i = arnFormat.indexOf('${');
  const j = arnFormat.indexOf('}', i);

  const prefix = arnFormat.substring(0, i);
  const varName = arnFormat.substring(i + 2, j);
  const suffix = arnFormat.substring(j + 1);

  const result: Expression[] = [];
  if (prefix !== '') {
    result.push(expr.lit(prefix));
  }
  result.push(expr.ident(`${refName}.${varName}`));
  if (suffix !== '') {
    result.push(expr.lit(suffix));
  }

  return result;
}
