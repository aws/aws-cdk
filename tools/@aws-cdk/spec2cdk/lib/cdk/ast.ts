import { SpecDatabase, Resource, Service } from '@aws-cdk/service-spec-types';
import { Module } from '@cdklabs/typewriter';
import { AugmentationsModule } from './augmentation-generator';
import { CannedMetricsModule } from './canned-metrics';
import { CDK_CORE, CONSTRUCTS, ModuleImportLocations } from './cdk';
import { ResourceClass } from './resource-class';

/**
 * A module containing a single resource
 */
export class ResourceModule extends Module {
  public constructor(public readonly service: string, public readonly resource: string) {
    super(`@aws-cdk/${service}/${resource}-l1`);
  }
}

/**
 * A module containing a service
 */
export class ServiceModule extends Module {
  public constructor(public readonly service: string, public readonly shortName: string) {
    super(`@aws-cdk/${service}`);
  }
}

export interface AstBuilderProps {
  readonly db: SpecDatabase;
  /**
   * Override the locations modules are imported from
   */
  readonly importLocations?: ModuleImportLocations;

  /**
   * Append a suffix at the end of generated names.
   */
  readonly nameSuffix?: string;
}

export class AstBuilder<T extends Module> {
  /**
   * Build a module for all resources in a service
   */
  public static forService(service: Service, props: AstBuilderProps): AstBuilder<ServiceModule> {
    const scope = new ServiceModule(service.name, service.shortName);
    const aug = new AugmentationsModule(props.db, service.name, props.importLocations?.cloudwatch);
    const metrics = CannedMetricsModule.forService(props.db, service);

    const ast = new AstBuilder(scope, props, aug, metrics);

    const resources = props.db.follow('hasResource', service);

    for (const link of resources) {
      ast.addResource(link.entity);
    }

    return ast;
  }

  /**
   * Build an module for a single resource
   */
  public static forResource(resource: Resource, props: AstBuilderProps): AstBuilder<ResourceModule> {
    const parts = resource.cloudFormationType.toLowerCase().split('::');
    const scope = new ResourceModule(parts[1], parts[2]);
    const aug = new AugmentationsModule(props.db, parts[1], props.importLocations?.cloudwatch);
    const metrics = CannedMetricsModule.forResource(props.db, resource);

    const ast = new AstBuilder(scope, props, aug, metrics);
    ast.addResource(resource);

    return ast;
  }

  public readonly db: SpecDatabase;
  /**
   * Map of CloudFormation resource name to generated class name
   */
  public readonly resources: Record<string, string> = {};
  private nameSuffix?: string;

  protected constructor(
    public readonly module: T,
    props: AstBuilderProps,
    public readonly augmentations?: AugmentationsModule,
    public readonly cannedMetrics?: CannedMetricsModule,
  ) {
    this.db = props.db;
    this.nameSuffix = props.nameSuffix;

    CDK_CORE.import(this.module, 'cdk', { fromLocation: props.importLocations?.core });
    CONSTRUCTS.import(this.module, 'constructs');
    CDK_CORE.helpers.import(this.module, 'cfn_parse', { fromLocation: props.importLocations?.coreHelpers });
  }

  public addResource(resource: Resource) {
    const resourceClass = new ResourceClass(this.module, this.db, resource, this.nameSuffix);
    this.resources[resource.cloudFormationType] = resourceClass.spec.name;

    resourceClass.build();

    this.augmentations?.augmentResource(resource, resourceClass);
  }
}
