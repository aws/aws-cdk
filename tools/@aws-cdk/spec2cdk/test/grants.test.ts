import { loadAwsServiceSpec } from '@aws-cdk/aws-service-spec';
import { SpecDatabase } from '@aws-cdk/service-spec-types';
import { TypeScriptRenderer } from '@cdklabs/typewriter';
import { GrantsModule } from '../lib/cdk/grants-module';

const renderer = new TypeScriptRenderer();
let db: SpecDatabase;

beforeAll(async () => {
  db = await loadAwsServiceSpec();
});

test('generates grants for methods with and without key actions', async () => {
  const config =JSON.stringify({
    resources: {
      Topic: {
        hasResourcePolicy: true,
        grants: {
          publish: {
            actions: ['sns:Publish'],
            keyActions: ['kms:Decrypt', 'kms:GenerateDataKey*'],
          },
          subscribe: {
            actions: ['sns:Subscribe'],
          },
        },
      },
    },
  });
  const service = db.lookup('service', 'name', 'equals', 'aws-sns').only();
  const module = await GrantsModule.forServiceFromString(db, service, config);

  if (module == null) {
    throw new Error('No grants module generated');
  }

  const rendered = renderer.render(module);
  expect(rendered).toMatchSnapshot();
});
