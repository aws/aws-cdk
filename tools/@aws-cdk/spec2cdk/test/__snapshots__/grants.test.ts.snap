// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asdas 1`] = `
"/* eslint-disable prettier/prettier, @stylistic/max-len */
import * as sns from "aws-cdk-lib/aws-sns";
import * as iam from "aws-cdk-lib/aws-iam";

export class TopicGrants {
  public readonly resource: sns.ITopicRef;

  public readonly encryptedResource?: iam.IEncryptedResource;

  public readonly policyResource?: iam.IResourceWithPolicy;

  public constructor(resource: sns.ITopicRef, encryptedResource?: iam.IEncryptedResource, policyResource?: iam.IResourceWithPolicy) {
    this.resource = resource;
    this.encryptedResource = encryptedResource;
    this.policyResource = policyResource;
  }

  public publish(grantee: iam.IGrantable): iam.Grant {
    const actions = ["sns:Publish"];
    this.encryptedResource?.grantOnKey(grantee, "kms:Decrypt", "kms:GenerateDataKey*");
    return (this.policyResource ? iam.Grant.addToPrincipalOrResource({
      "actions": actions,
      "grantee": grantee,
      "resourceArns": [this.resource.topicRef.topicArn],
      "resource": this.policyResource
    }) : iam.Grant.addToPrincipal({
      "actions": actions,
      "grantee": grantee,
      "resourceArns": [this.resource.topicRef.topicArn]
    }));
  }

  public subscribe(grantee: iam.IGrantable): iam.Grant {
    const actions = ["sns:Subscribe"];
    return (this.policyResource ? iam.Grant.addToPrincipalOrResource({
      "actions": actions,
      "grantee": grantee,
      "resourceArns": [this.resource.topicRef.topicArn],
      "resource": this.policyResource
    }) : iam.Grant.addToPrincipal({
      "actions": actions,
      "grantee": grantee,
      "resourceArns": [this.resource.topicRef.topicArn]
    }));
  }
}"
`;
